core/wrappers/tasks.py

```diff
@@ -0,0 +1,132 @@
+        obs_delay_range=range(0, 8),
+        act_delay_range=range(0, 2),
+        instant_rewards: bool = True,
+    ):
+        super().__init__(
+            env, observation_space=env.observation_space, action_space=env.action_space
+        )
+
+        self.act_delay_range = act_delay_range
+        self.obs_delay_range = obs_delay_range
+        self.instant_rewards = instant_rewards
+
+        self.past_actions = deque(
+            maxlen=obs_delay_range.stop + act_delay_range.stop - 1
+        )
+        self.past_observations = deque(maxlen=obs_delay_range.stop)
+        self.arrival_times_actions = deque(maxlen=act_delay_range.stop)
+        self.arrival_times_observations = deque(maxlen=obs_delay_range.stop)
+
+        self.t = 0
+        self.current_action = None
+
+    def reset(self) -> VecEnvObs:
+        num_wrapped_envs = self.venv.num_envs
+        first_observation = self.venv.reset()
+
+        # fill up buffers
+        self.t = -(self.obs_delay_range.stop + self.act_delay_range.stop)
+        while self.t < 0:
+            self.send_action(
+                np.zeros(
+                    (num_wrapped_envs, *self.venv.action_space.shape),
+                    dtype=np.float32,
+                )
+            )
+            self.send_observation(
+                (
+                    first_observation,
+                    np.zeros((num_wrapped_envs,), dtype=np.float32),
+                    np.zeros((num_wrapped_envs,), dtype=np.bool_),
+                    [{}] * num_wrapped_envs,
+                    0,
+                )
+            )
+            self.t += 1
+
+        assert self.t == 0
+        received_observation, *_ = self.receive_observation()
+        return received_observation
+
+    def step_async(self, action: np.ndarray) -> None:
+        # at the brain
+        self.send_action(action)
+
+    def step_wait(self) -> VecEnvStepReturn:
+        num_wrapped_envs = self.venv.num_envs
+
+        # at the remote actor, for all ifs
+        if self.t < self.act_delay_range.stop:
+            # do nothing until the brain's first actions arrive at the remote actor
+            self.receive_action()
+            aux = (
+                np.zeros((num_wrapped_envs,), dtype=np.float32),
+                np.zeros((num_wrapped_envs,), dtype=np.bool_),
+                [{}] * num_wrapped_envs,
+            )
+        else:
+            action_delay = self.receive_action()
+
+            self.venv.step_async(
+                self.current_action
+            )  # we make sure the wrapped venv knows the action
+            m, *aux = self.venv.step_wait()  # we get the results from the wrapped venv
+
+            self.send_observation((m, *aux, action_delay))
+
+        # at the brain again
+        m, *delayed_aux = self.receive_observation()
+        aux = aux if self.instant_rewards else delayed_aux
+        self.t += 1
+        return m, *aux
+
+    def send_action(self, action):
+        # at the brain
```

Comment:
Haven't run the code yet so im not sure how feasible this is but would be good to have this implemented alongside domain randomization code? Meaning the obs delays and action delay feature  are apart of dr. Don't think it changes much other than the fact that it might easier to explain to someone using the code base for the first time.

Reply:
I think it's essentially different, because DR involves much more the Isaac Sim part of the pipeline, while this is purely the STB3 part...

Does that make sense?